### ISSUE TITLE
Public key authentication threading issue with openssl in FIPS mode

### ISSUE THREAD

**Issue by jonathancrowe** [libssh2/libssh2#188]
When establishing an SSH or SFTP session using libssh2, the `libssh2_userauth_publickey_fromfile` function may be used to pass a private key for user authentication. When linked against openssl, the `_libssh2_pub_priv_keyfile` function in src/openssl.c is called. That method checks to see if the "des" cipher can be retrieved, and if it cannot, it calls `OpenSSL_add_all_ciphers`. This also occurs in the `_libssh2_pub_priv_keyfilememory` function. The code block looks the same in both:

```
#if OPENSSL_VERSION_NUMBER < 0x10100000L
    if (!EVP_get_cipherbyname("des")) {
        /* If this cipher isn't loaded it's a pretty good indication that none
         * are.  I have *NO DOUBT* that there's a better way to deal with this
         * ($#&%#$(%$#( Someone buy me an OpenSSL manual and I'll read up on
         * it.
         */
        OpenSSL_add_all_ciphers();
    }
#endif
```

The problem is that `OpenSSL_add_all_ciphers` is not a thread-safe call. If multiple threads try to establish a connection in the same way at the same time, the application could crash. However, this is made even worse when openssl is operating in FIPS mode, because the "des" cipher will never be loaded in FIPS mode. Therefore, if in FIPS mode, there is a good chance that a process will crash if it is establishing connections using libssh2 in multiple threads.

I have attached a sample C file, [libssh2_keyauth.c.txt](https://github.com/libssh2/libssh2/files/1062303/libssh2_keyauth.c.txt), that can be used to reproduce this issue. I am building on RHEL7.2, using libssh2-1.4.3-10.el7_2.1, but I see the `OpenSSL_add_all_ciphers` call is still present in the latest master branch. I compiled using:
```gcc libssh2_keyauth.c -lssh2 -lssl -lcrypto -pthread```

I then ran the application using:
```./a.out host user /path/to/private/key 4```
where 4 is the number of threads. I was able to reproduce the crash after a few tries, but it may vary for others. Of course, a proper host name, username, and private key file must be passed as arguments, and the server must accept the key the client is presenting.

Below is the stack trace captured from one of the aborts I saw.

```
#0  0x00007ffff6fb35f7 in raise () from /lib64/libc.so.6
#1  0x00007ffff6fb4ce8 in abort () from /lib64/libc.so.6
#2  0x00007ffff6ff3327 in __libc_message () from /lib64/libc.so.6
#3  0x00007ffff6ffb053 in _int_free () from /lib64/libc.so.6
#4  0x00007ffff75c54ed in CRYPTO_free () from /lib64/libcrypto.so.10
#5  0x00007ffff75c8001 in OBJ_NAME_add () from /lib64/libcrypto.so.10
#6  0x00007ffff764a655 in EVP_add_cipher () from /lib64/libcrypto.so.10
#7  0x00007ffff764e9cb in OpenSSL_add_all_ciphers () from /lib64/libcrypto.so.10
#8  0x00007ffff7bd0255 in _libssh2_pub_priv_keyfile () from /lib64/libssh2.so.1
#9  0x00007ffff7bcb84e in libssh2_userauth_publickey_fromfile_ex () from /lib64/libssh2.so.1
#10 0x00000000004017c5 in connect_and_authenticate (host=0x7fffffffe5cf "host", user=0x7fffffffe5db "user", key_file=0x7fffffffe5e3 "/path/to/private/key") at libssh2_keyauth.c:74
#11 0x000000000040185d in thread_connect_and_authenticate (p=0x6208f0) at libssh2_keyauth.c:101
#12 0x00007ffff7347dc5 in start_thread () from /lib64/libpthread.so.0
#13 0x00007ffff7074ced in clone () from /lib64/libc.so.6
```

I can think of a couple solutions to this problem. While the "des" cipher is not loaded when openssl is in FIPS mode, the "des3" cipher is. Changing the check to get the "des3" cipher instead would work around the issue in FIPS mode, and it is also likely that the "des3" cipher is loaded by most applications in the same way that the "des" cipher is loaded.

Another solution would be to replace the entire code block that gets the "des" cipher and calls `OpenSSL_add_all_ciphers` with a call to `_libssh2_init_if_needed`. This function is used elsewhere in openssl.c, and has the benefit of checking if `libssh2_init` has been called already, which would be the common case for well-written multi-threaded applications using libssh2.

Another solution would be to write a separate code path for FIPS mode to use a different cipher lookup, such as "des3". This has the benefit of not changing the past behavior of using "des" as the lookup cipher, while supporting applications running with FIPS mode openssl.

Finally, a solution could be to remove the block of code entirely. This would require `libssh2_init` and/or `OpenSSL_add_all_ciphers` to be invoked by the developer using libssh2 ahead of time, but would eliminate the thread safety issue entirely.

**Comment by willco007** [comment]
This call should be moved into libssh2_crypto_init()

**Comment by willco007** [comment]
To make this work correctly with the `OPENSSL_VERSION_NUMBER < 0x10100000L` check we would need to formalize libssh2_crypto_init() into a function (like _libssh2_init_aes_ctr already is) which I'd be OK with.

**Comment by willco007** [comment]
@jonathancrowe could you try PR #189 with your test app and see if it resolves the issue for you?

**Comment by jonathancrowe** [comment]
Thank you for the quick response, @willco007! I rebuilt my test application to use the master branch of this project, and was able to reproduce the crash fairly often. I pulled down PR #189 and rebuilt my test application with those changes, and was unable to reproduce the crash after many attempts.

I did want to point out, though, that the same code change that was in `a69890c` should be made for the `_libssh2_pub_priv_keyfilememory` function in src/openssl.c, as well. That function would be used if my application used the `userauth_publickey_frommemory` function instead of the `userauth_publickey_fromfile` function.

Again, I appreciate how quickly you responded to this issue!
